### PR TITLE
issue: 1284069 Introduce plugin/plugout support

### PR DESCRIPTION
**Pull request by igor-ivanov**

### CODE REVIEW

File: src/vma/event/event_handler_manager.cpp [hunk 1]
```diff
@@ -409,7 +409,7 @@ void event_handler_manager::update_epfd(int fd, int operation, int events)
 	BULLSEYE_EXCLUDE_BLOCK_START
 	if (orig_os_api.epoll_ctl(m_epfd, operation, fd, &ev) < 0) {
 		const char* operation_str[] = {"", "ADD", "DEL", "MOD"};
-		evh_logerr("epoll_ctl(%d, %s, fd=%d) failed (errno=%d %m)", 
+		evh_logdbg("epoll_ctl(%d, %s, fd=%d) failed (errno=%d %m)",
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/event/event_handler_manager.cpp]: updated in the same commit

File: src/vma/dev/net_device_table_mgr.cpp [hunk 2]
```diff
@@ -169,6 +170,7 @@ void net_device_table_mgr::update_tbl()
 	char nl_res[8096];
 	static int _seq = 0;
 	net_device_val* p_net_device_val;
+	std::vector<int> cache_objs;
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/net_device_table_mgr.cpp]: cache_objects allows to store current list of ndevs and basing this information add new or remove unused.

File: src/vma/dev/net_device_table_mgr.cpp [hunk 3]
```diff
@@ -196,6 +198,16 @@ void net_device_table_mgr::update_tbl()
 	}
 
 	m_lock.lock();
+
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/net_device_table_mgr.cpp]: It is supposed that update_tbl() will help during migration

File: src/vma/dev/ring_bond.h [hunk 4]
```diff
@@ -172,7 +171,7 @@ class ring_bond_netvsc : public ring_bond
 protected:
 	virtual void slave_create(int if_index);
 
-private:
+public:
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/ring_bond.h]: they are used  (for example https://github.com/Mellanox/libvma/pull/607/files#diff-9190366b0d2117456bd385aac29fc2e7R183)

File: src/vma/dev/ib_ctx_handler.cpp [hunk 5]
```diff
@@ -277,19 +274,17 @@ uint32_t ib_ctx_handler::mem_reg(void *addr, size_t length, uint64_t access)
 
 void ib_ctx_handler::mem_dereg(uint32_t lkey)
 {
-	if (!is_removed()) {
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/ib_ctx_handler.cpp]: it is achronim as far in current implementation RTM_NEWLINK event is used that comes before IBV_EVENT_DEVICE_FATAL.

File: src/vma/dev/allocator.cpp [hunk 6]
```diff
@@ -291,3 +284,35 @@ bool vma_allocator::register_memory(size_t size, ib_ctx_handler *p_ib_ctx_h,
 
 	return true;
 }
+
+void vma_allocator::deregister_memory(ib_ctx_handler *p_ib_ctx_h)
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/allocator.cpp]: it should be used in case plugout to cleanup

File: src/vma/dev/ib_ctx_handler_collection.cpp [hunk 7]
```diff
@@ -85,12 +86,13 @@ void ib_ctx_handler_collection::update_tbl()
 	ib_ctx_handler * p_ib_ctx_handler = NULL;
 	int num_devices = 0;
 	int i;
+	std::vector<struct ibv_device*> cache_objs;
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/ib_ctx_handler_collection.cpp]: no reason

File: src/vma/dev/net_device_val.cpp [hunk 8]
```diff
@@ -873,13 +856,89 @@ bool net_device_val::update_active_slaves() {
...
+				m_slaves[i]->active = true;
+			}
+		}
+
+		m_slaves[i]->p_ib_ctx = g_p_ib_ctx_handler_collection->get_ib_ctx(base_ifname);
+		m_slaves[i]->port_num = get_port_from_ifname(base_ifname);
+	}
+
+	/* restart if status changed */
+	if (changed) {
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/net_device_val.cpp]: to be symmetric with the same function active bond.

File: src/vma/dev/ring_bond.cpp [hunk 9]
```diff
@@ -293,6 +415,8 @@ bool ring_bond::get_hw_dummy_send_support(ring_user_id_t id, vma_ibv_send_wr* p_
 			return false;
 		}
 	}
+
+	return false;
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/ring_bond.cpp]: updated at the same commit

File: src/vma/dev/ring_bond.cpp [hunk 10]
```diff
@@ -218,6 +324,7 @@ void ring_bond::restart(ring_resource_creation_info_t* p_ring_info) {
 
 void ring_bond::adapt_cq_moderation()
 {
+	auto_unlocker lock(m_lock_ring_rx);
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/ring_bond.cpp]: updated

File: src/vma/dev/ring_allocation_logic.cpp [hunk 11]
```diff
@@ -84,6 +84,9 @@ uint64_t ring_allocation_logic::calc_res_key_by_logic()
 		if (safe_mce_sys().tcp_ctl_thread > CTL_THREAD_DISABLE)
 			res_key = 1;
 		break;
+	case RING_LOGIC_PER_IP:
+		res_key = m_res_key.get_user_id_key();
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/ring_allocation_logic.cpp]: updated

File: src/vma/dev/net_device_table_mgr.cpp [hunk 12]
```diff
@@ -297,20 +337,54 @@ net_device_val* net_device_table_mgr::get_net_device_val(in_addr_t local_addr)
...
+			if (if_index == slaves[i]->if_index) {
+				goto out;
+			}
+		}
+		/* Check if interface is new slave */
+		char if_name[IFNAMSIZ] = {0};
+		char sys_path[256] = {0};
+		int ret = 0;
+		if (if_indextoname(if_index, if_name)) {
+			ret = snprintf(sys_path, sizeof(sys_path), NETVSC_DEVICE_UPPER_FILE, if_name, net_dev->get_ifname());
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/net_device_table_mgr.cpp]: plugin: https://github.com/Mellanox/libvma/pull/607/commits/2911cb9f1b9eeb67b1671cb00c7397843a22ee1d#diff-b353ac0905c8a33b9daa213d93de2d07R646

[the remark this answers is not in the document]
Comment (liranoz12) [reply src/vma/dev/net_device_table_mgr.cpp]: There is no need for it - use can find the netdev index in the following path :+1: 
/sys/class/net/enP2p0s2/upper_eth1/ifindex

File: src/vma/dev/net_device_val.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (liranoz12) [reply src/vma/dev/net_device_val.cpp]: I do not think consistent use should be our primary goal.

File: src/vma/dev/net_device_table_mgr.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (liranoz12) [reply src/vma/dev/net_device_table_mgr.cpp]: so we should add this logic during at the next phase

File: src/vma/dev/allocator.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (liranoz12) [reply src/vma/dev/allocator.cpp]: if currently there is no use for the input paramter there is no reason to push it.

File: src/vma/dev/ring_bond.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (liranoz12) [reply src/vma/dev/ring_bond.h]: The implementation of ring bond::restart is too long. we cant push it like this.
ring bond netvsc inherits from ring_bond, why cant we implement a new virtual restart function ? 
why do we need to have a pointer to the vf_ring and to m_tap_ring? the root cause for the new ring_slave type is to avoid those spespic calls and use virtual function instead. - we should impelemt restart function for each ring_salve.

File: src/vma/dev/ib_ctx_handler.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (liranoz12) [reply src/vma/dev/ib_ctx_handler.cpp]: are we sure that this check is redundant ? if not please don't remove it,.

File: src/vma/dev/allocator.cpp [hunk 13]
```diff
@@ -93,30 +83,32 @@ void* vma_allocator::alloc_and_reg_mr(size_t size, ib_ctx_handler *p_ib_ctx_h)
...
+			m_mem_alloc_type = ALLOC_TYPE_HUGEPAGES;
 			break;
 		}
 	// fallthrough
 	case ALLOC_TYPE_CONTIG:
-		if (m_is_contig_alloc &&
-				g_p_ib_ctx_handler_collection->get_ib_cxt_list()) {
-			if (!register_memory(size, p_ib_ctx_h, m_contig_access_mr)) {
+#ifdef VMA_IBV_ACCESS_ALLOCATE_MR
+		if (!safe_mce_sys().cpuid_hv() && g_p_ib_ctx_handler_collection->get_ib_cxt_list()) {
```
Comment (liranoz12) [inline src/vma/dev/allocator.cpp]: Why not use safe_mce_sys().is_hypervisor ?

File: src/vma/event/event_handler_manager.cpp [hunk 14]
```diff
@@ -407,9 +407,10 @@ void event_handler_manager::update_epfd(int fd, int operation, int events)
 	ev.events = events;
 	ev.data.fd = fd;
 	BULLSEYE_EXCLUDE_BLOCK_START
-	if (orig_os_api.epoll_ctl(m_epfd, operation, fd, &ev) < 0) {
+	if ((orig_os_api.epoll_ctl(m_epfd, operation, fd, &ev) < 0) &&
+		(!(errno == ENOENT || errno == EBADF))) {
```
Comment (liranoz12) [inline src/vma/event/event_handler_manager.cpp]: What is the reason for this change ? we should resolve the problem not modify the error message

Comment (liranoz12) [reply src/vma/event/event_handler_manager.cpp]: we should suppress the error in priv_unregister_ibverbs_events and no in update_epfd

File: src/vma/dev/allocator.cpp (hunk 13 above)
Comment (igor-ivanov) [reply src/vma/dev/allocator.cpp]: no reason. will update

File: src/vma/event/event_handler_manager.cpp (hunk 14 above)
Comment (igor-ivanov) [reply src/vma/event/event_handler_manager.cpp]: The same technique is used as for identical issue at https://github.com/Mellanox/libvma/blob/master/src/vma/dev/net_device_table_mgr.cpp#L455
https://github.com/Mellanox/libvma/blob/master/src/vma/iomux/epfd_info.cpp#L704

**Comment by liranoz12** [comment]
duplicate by #632
